File: climchg.m
function varargout = climchg(vTemp, vTime, sIndex, varargin)
% this will asses cirtain climate data sets and so that correct
% understanding of what the climatae is doing can be asertained.
%varagin allows more inputs that you put anything in



    switch sIndex
        case 'slope to target'
            %find target date
            dTarget = varargin{1};
            %dDate = find(vTime==dTarget);
            iElNum = find(abs(vTime - dTarget) == min(abs(vTime - dTarget)), 1);
            %cut off data after target date
            vTiCut = vTime(1:iElNum);
            vTempCut = vTemp(1:iElNum);
            
            %preallocating memory
            vStartTime = zeros(length(vTiCut) - 4, 1);
            vSlope = zeros(length(vTiCut) - 4, 1);
            vUB = zeros(length(vTiCut) - 4, 1);
            vLB = zeros(length(vTiCut) - 4, 1);
            
            for i = 1:length(vTiCut) - 4
                [xData, yData] = prepareCurveData( vTiCut(i:end), vTempCut(i:end));
                
                ft = fittype('poly1');
    
                fitresult = fit(xData, yData, ft);
    
                %get the data out of variables 
                vCoeffs = coeffvalues(fitresult);
                mConfInt =  confint(fitresult);
    
                %assign data to output vectors
    
                vStartTime(i) = vTiCut(i);
                vSlope(i) = vCoeffs(1) * 10;
                vLB(i) = mConfInt(1,1) * 10;
                vUB(i) = mConfInt(2,1) * 10;
            end
        
            % set up variables for plotting
            %vTimePlot = vTiCut(1:end-4);
        
            
            %plot data
            plot(vStartTime, vSlope, 'blue')
            hold on;
            plot(vStartTime, vUB, 'red')
            plot(vStartTime, vLB, 'red')
            hold off;

            %change data labels
                xlabel('Date', 'FontSize', 10);
                sTitle = sprintf('slope to target, %.0f', dTarget);
                title(sTitle);            
                dIfVariable = vTime(2) - vTime(1);
                    if dIfVariable == 1
                        sYLabel = sprintf('Slope from date to %.0f (C° per decade)\n annual data',...
                            dTarget);
                    else 
                        sYLabel = sprintf('Slope from date to %.0f (C° per decade)\n monthly data',...
                            dTarget);                
                    end
                ylabel(sYLabel, 'FontSize', 10);


            %below are the outputs for this case, make sure you call for 
            %4 variables
            varargout{1} = vStartTime;
            varargout{2} = vSlope;
            varargout{3} = vUB;
            varargout{4} = vLB;
    
        case 'slope of interval'
            %find end target 
            dInterval = varargin{1};
            
            %adjust intervals accept month or year data
            if vTime(2) - vTime(1) > 0.9
                iWin = dInterval;
            elseif vTime(2) - vTime(1) < 0.9
                iWin = dInterval*12;
            end

            %set up place holders for variable made in folowing loop
            vSlope = zeros(length(vTime) - iWin + 1, 1);
            vUB = zeros(length(vTime) - iWin + 1, 1);
            vLB = zeros(length(vTime) - iWin + 1, 1);
            vEndTime = zeros(length(vTime) - iWin + 1, 1);
            vEndCell = zeros(length(vTime) - iWin + 1, 1);

            %Create slope for each interval
            for i = 1:length(vTime) - iWin + 1

                %Cut off after target date
                vTiCut = vTime(i:iWin + i - 1);
                vTempCut = vTemp(i:iWin + i - 1);

                %prepare the data for fit
                [vxData, vyData] = prepareCurveData(vTiCut, vTempCut);

                %Fit type defined as first order polynomial
                ft = fittype( 'poly1' );

                %Fit model to data
                [fitresult, gof] = fit( vxData, vyData, ft);
                %extract data
                vCoeffs = coeffvalues(fitresult);
                mConfInt = confint(fitresult);

                %assign data to output vectors                 
                vSlope(i) = vCoeffs(1) * 10;          
                vUB(i) = mConfInt(2,1) * 10;
                vLB(i) = mConfInt(1,1) * 10;
                %Set up the end date for each interval
                vEndCell(i) = iWin + i - 1;
                vEndTime(i) = vTime(vEndCell(i));


            end

            %set up output variables... this order is important for calling
            %them in the function 
            varargout{1} = vEndTime;
            varargout{2} = vSlope;
            varargout{3} = vUB;
            varargout{4} = vLB;

            %plot (slope) vs (end date) and Confedence interval lines
            plot(vEndTime, vSlope, 'blue')
            hold on;
            plot(vEndTime, vUB, 'red')
            plot(vEndTime, vLB, 'red')
            hold off;
            %change data labels
                xlabel('Date', 'FontSize', 10);
                sTitle = sprintf('slope of interval size %.0f',dInterval);
                title(sTitle);            
                dIfVariable = vTime(2) - vTime(1);
                    if dIfVariable == 1
                        sYLabel = sprintf('Slope of interval size %.0f, C° per decade\n (yearly data)',...
                            dInterval);
                    else 
                        sYLabel = sprintf('Slope of interval size %.0f, C° per decade\n (monthly data)',...
                            dInterval);                
                    end
                ylabel(sYLabel, 'FontSize', 10);


        case 'trend since 1970'
            %find 1970, calculate till end
            dElNum = find(abs(vTime - 1970) == min(abs(vTime - 1970)), 1);
            vTiCut = vTime(dElNum:end);
            vTempCut = vTemp(dElNum:end);
            
            % set up curve fitting function
            [vxData, vyData] = prepareCurveData(vTiCut, vTempCut);
            %set up linear fit
            ft = fittype( 'poly1' );
            %fit model to the data
            [fitresult, gof] = fit( vxData, vyData, ft );

            %plot data and fit
            figure( "Name", 'Trend Since 1970');
                plot(fitresult, vxData, vyData, 'predobs');
            %change data labels
                xlabel('Date', 'FontSize', 10);
                sTitle = sprintf('Trend Since 1970');
                title(sTitle);            
                dIfVariable = vTime(2) - vTime(1);
                    if dIfVariable == 1
                        sYLabel = sprintf('Average Annual Temperature');
                    else 
                        sYLabel = sprintf('Average Monthly Temperature');                
                    end
                ylabel(sYLabel, 'FontSize', 10);

            
            %fitting stats
            vCoeffs = coeffvalues(fitresult);
            mConfInt = confint(fitresult);     

            dSlope = vCoeffs(1);
            dIntercept = vCoeffs(2);
            vUB = mConfInt(2,1);
            vLB = mConfInt(1,1);
            
            %fprintf('slope is %.3G\n --> 95 percent confindence interval of [%.3G, %.3G]\n',dSlope,vUB,vLB);
            
            %below are the outputs for this case, make sure you call for 
            %4 variables
            varargout{1} = dIntercept;
            varargout{2} = dSlope;
            varargout{3} = vUB;
            varargout{4} = vLB;

        otherwise           
            error('Your sIndex value is not supported. Try one of the following: -slope to target-slope of interval-trend since 1970-');
                        
         
    end

 
end

File: climchg.py
# -*- coding: utf-8 -*-
"""
Created on Fri Oct  6 08:50:27 2023

@author: duckm
"""
#A python equivelant to the matlab climchg.m program
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit

plt.rcParams['figure.dpi'] = 300

def climchg(vTemp, vTime, sIndex, *args):
    if sIndex == 'slope to target':
        # Extract the target date from the arguments
        dTarget = args[0]
        # Find the index of the target date in vTime
        iElNum = np.argmin(np.abs(vTime - dTarget))
        # Cut the data up to the target date
        vTiCut = vTime[:iElNum + 1]
        vTempCut = vTemp[:iElNum + 1]

        # Initialize arrays to store results
        vStartTime = np.zeros(len(vTiCut) - 4)
        vSlope = np.zeros(len(vTiCut) - 4)
        vUB = np.zeros(len(vTiCut) - 4)
        vLB = np.zeros(len(vTiCut) - 4)

        # Loop to calculate slope for each interval
        for i in range(len(vTiCut) - 4):
            xData = vTiCut[i:]
            yData = vTempCut[i:]
            
            # Define a linear function
            def linear_func(x, a, b):
                return a * x + b
            
            # Fit the data to the linear function
            popt, _ = curve_fit(linear_func, xData, yData)
            vStartTime[i] = vTiCut[i]
            vSlope[i] = popt[0] * 10
            vLB[i] = popt[0] * 10 - 1.96 * np.std(yData) / np.sqrt(len(yData))
            vUB[i] = popt[0] * 10 + 1.96 * np.std(yData) / np.sqrt(len(yData))

        # Plot the results
        plt.plot(vStartTime, vSlope, 'blue')
        plt.plot(vStartTime, vUB, 'red')
        plt.plot(vStartTime, vLB, 'red')
        plt.xlabel('Date', fontsize=10)
        sTitle = f'slope to target, {dTarget:.0f}'
        plt.title(sTitle)
        dIfVariable = vTime[1] - vTime[0]
        if dIfVariable == 1:
            sYLabel = f'Slope from date to {dTarget:.0f} (C° per decade)\n annual data'
        else:
            sYLabel = f'Slope from date to {dTarget:.0f} (C° per decade)\n monthly data'
        plt.ylabel(sYLabel, fontsize=10)
        
        # Return the results
        return vStartTime, vSlope, vUB, vLB

    elif sIndex == 'slope of interval':
        # Extract the interval size from the arguments
        dInterval = args[0]

        # Adjust interval based on data frequency (monthly or yearly)
        if vTime[1] - vTime[0] > 0.9:
            iWin = int(dInterval)
        else:
            iWin = int(dInterval * 12)

        # Initialize arrays to store results
        vSlope = np.zeros(len(vTime) - iWin + 1)
        vUB = np.zeros(len(vTime) - iWin + 1)
        vLB = np.zeros(len(vTime) - iWin + 1)
        vEndTime = np.zeros(len(vTime) - iWin + 1)
        vEndCell = np.zeros(len(vTime) - iWin + 1)

        # Loop to calculate slope for each interval
        for i in range(len(vTime) - iWin + 1):
            vTiCut = vTime[i:i + iWin]
            vTempCut = vTemp[i:i + iWin]

            # Define a linear function
            def linear_func(x, a, b):
                return a * x + b
            
            # Fit the data to the linear function
            popt, _ = curve_fit(linear_func, vTiCut, vTempCut)
            vSlope[i] = popt[0] * 10
            vLB[i] = popt[0] * 10 - 1.96 * np.std(vTempCut) / np.sqrt(len(vTempCut))
            vUB[i] = popt[0] * 10 + 1.96 * np.std(vTempCut) / np.sqrt(len(vTempCut))
            vEndCell[i] = iWin + i - 1
            vEndTime[i] = vTime[int(vEndCell[i])]

        # Plot the results
        plt.plot(vEndTime, vSlope, 'blue')
        plt.plot(vEndTime, vUB, 'red')
        plt.plot(vEndTime, vLB, 'red')
        plt.xlabel('Date', fontsize=10)
        sTitle = f'slope of interval size {dInterval:.0f}'
        plt.title(sTitle)
        dIfVariable = vTime[1] - vTime[0]
        if dIfVariable == 1:
            sYLabel = f'Slope of interval size {dInterval:.0f}, C° per decade\n (yearly data)'
        else:
            sYLabel = f'Slope of interval size {dInterval:.0f}, C° per decade\n (monthly data)'
        plt.ylabel(sYLabel, fontsize=10)

        # Return the results
        return vEndTime, vSlope, vUB, vLB

    elif sIndex == 'trend since 1970':
        # Find the index corresponding to 1970
        dElNum = np.argmin(np.abs(vTime - 1970))
        # Cut the data from 1970 to the end
        vTiCut = vTime[dElNum:]
        vTempCut = vTemp[dElNum:]

        # Define a linear function
        def linear_func(x, a, b):
            return a * x + b
        
        # Fit the data to the linear function
        popt, _ = curve_fit(linear_func, vTiCut, vTempCut)

        # Plot the data and fitted line
        plt.plot(vTiCut, vTempCut, 'b.')
        plt.plot(vTiCut, linear_func(vTiCut, *popt), 'r-')
        plt.xlabel('Date', fontsize=10)
        plt.title('Trend Since 1970')
        dIfVariable = vTime[1] - vTime[0]
        if dIfVariable == 1:
            sYLabel = 'Average Annual Temperature'
        else:
            sYLabel = 'Average Monthly Temperature'
        plt.ylabel(sYLabel, fontsize=10)

        # Extract the fitting parameters
        dSlope = popt[0] * 10
        dIntercept = popt[1]
        vUB = 1.96 * np.std(vTempCut) / np.sqrt(len(vTempCut))
        vLB = -vUB
        
        # Return the results
        return dIntercept, dSlope, vUB, vLB

    else:
        raise ValueError('Invalid sIndex value. Try one of the following: "slope to target", "slope of interval", "trend since 1970"')
